File: bms/go/geo/geo_r2_Rect_ExpandedByMargin/geo_r2_Rect_ExpandedByMargin.go
package test

func (r geo_r2_Rect) ExpandedByMargin(margin float64) geo_r2_Rect {
	return r.Expanded(geo_r2_Point{margin, margin})
}

func (r geo_r2_Rect) Expanded(margin geo_r2_Point) geo_r2_Rect {
	xx := r.X.Expanded(margin.X)
	yy := r.Y.Expanded(margin.Y)
	if xx.IsEmpty() || yy.IsEmpty() {
		return geo_r2_EmptyRect()
	}
	return geo_r2_Rect{xx, yy}
}

func (i geo_r1_Interval) Expanded(margin float64) geo_r1_Interval {
	if i.IsEmpty() {
		return i
	}
	return geo_r1_Interval{i.Lo - margin, i.Hi + margin}
}

func (i geo_r1_Interval) IsEmpty() bool { return i.Lo > i.Hi }

func geo_r2_EmptyRect() geo_r2_Rect {
	return geo_r2_Rect{geo_r1_EmptyInterval(), geo_r1_EmptyInterval()}
}

func geo_r1_EmptyInterval() geo_r1_Interval { return geo_r1_Interval{1, 0} }

type geo_r2_Rect struct {
	X, Y geo_r1_Interval
}

type geo_r1_Interval struct {
	Lo, Hi float64
}

type geo_r2_Point struct {
	X, Y float64
}

File: translations_only/fail/go/geo/geo_r2_Rect_ExpandedByMargin/mistral/BaseRepair/geo_r2_Rect_ExpandedByMargin.rs

struct GeoR2Rect {
    x: GeoR1Interval,
    y: GeoR1Interval,
}

struct GeoR1Interval {
    lo: f64,
    hi: f64,
}

struct GeoR2Point {
    x: f64,
    y: f64,
}

impl GeoR2Rect {
    fn expanded_by_margin(&self, margin: f64) -> GeoR2Rect {
        self.expanded(GeoR2Point {
            x: margin,
            y: margin,
        })
    }

    fn expanded(&self, margin: GeoR2Point) -> GeoR2Rect {
        let xx = self.x.expanded(margin.x);
        let yy = self.y.expanded(margin.y);
        if xx.is_empty() || yy.is_empty() {
            GeoR2EmptyRect
        } else {
            GeoR2Rect { x: xx, y: yy }
        }
    }
}

impl GeoR1Interval {
    fn expanded(&self, margin: f64) -> GeoR1Interval {
        if self.is_empty() {
            GeoR1EmptyInterval
        } else {
            GeoR1Interval {
                lo: self.lo - margin,
                hi: self.hi + margin,
            }
        }
    }

    fn is_empty(&self) -> bool {
        self.lo > self.hi
    }
}

const GeoR2EmptyRect: GeoR2Rect = GeoR2Rect {
    x: GeoR1EmptyInterval,
    y: GeoR1EmptyInterval,
};

const GeoR1EmptyInterval: GeoR1Interval = GeoR1Interval { lo: 1.0, hi: 0.0 };
